numpyro/mcmc.py

```diff
@@ -0,0 +1,111 @@
+        unflatten_fn = pack_fn
+        m_inv_init = np.ones(np.shape(z_flat))
+        rng, rng_momentum = random.split(rng)
+        momentum_generator = partial(_sample_momentum, m_inv_init, pack_fn, rng_momentum)
+        find_initial_ss = partial(find_reasonable_step_size, potential_fn, kinetic_fn, momentum_generator,
+                                  m_inv_init, init_samples)
+
+        wa_init, wa_update = warmup_adapter(num_warmup_steps,
+                                            find_reasonable_step_size=find_initial_ss,
+                                            adapt_step_size=adapt_step_size,
+                                            adapt_mass_matrix=adapt_mass_matrix,
+                                            diag_mass=diag_mass,
+                                            target_accept_prob=target_accept_prob)
+
+        wa_state = wa_init(step_size, mass_matrix_size=np.size(z_flat))
+        r = _sample_momentum(wa_state.inverse_mass_matrix, pack_fn, rng)
+        vv_state = vv_init(init_samples, r)
+
+        def body_fn(t, args):
+            vv_state, wa_state, rng = args
+            rng, rng_momentum, rng_transition = random.split(rng, 3)
+            r = _sample_momentum(wa_state.inverse_mass_matrix, pack_fn, rng_momentum)
+            vv_state = vv_state.update(r=r)
+            num_steps = _get_num_steps(wa_state.step_size, trajectory_length)
+            accept_prob, vv_state_new = _next(num_steps, wa_state.step_size, wa_state.inverse_mass_matrix, vv_state)
+            transition = random.bernoulli(rng_transition, accept_prob)
+            vv_state = cond(transition,
+                            vv_state_new, lambda state: state,
+                            vv_state, lambda state: state)
+            z_flat, _ = ravel_pytree(vv_state.z)
+            wa_state = wa_update(t, accept_prob, z_flat, wa_state)
+            return vv_state, wa_state, rng
+
+        vv_state, wa_state, rng = fori_loop(0, num_warmup_steps, body_fn, (vv_state, wa_state, rng))
+        num_steps = _get_num_steps(wa_state.step_size, trajectory_length)
+        return HMCState(vv_state.z, vv_state.z_grad, vv_state.potential_energy, num_steps,
+                        wa_state.step_size, wa_state.inverse_mass_matrix, rng)
+
+    def _next(num_steps, step_size, inverse_mass_matrix, vv_state):
+        vv_state_new = fori_loop(0, num_steps,
+                                 lambda i, val: vv_update(step_size, inverse_mass_matrix, val),
+                                 vv_state)
+        energy_old = vv_state.potential_energy + kinetic_fn(vv_state.r, inverse_mass_matrix)
+        energy_new = vv_state_new.potential_energy + kinetic_fn(vv_state_new.r, inverse_mass_matrix)
+        delta_energy = energy_new - energy_old
+        delta_energy = np.where(np.isnan(delta_energy), np.inf, delta_energy)
+        accept_prob = np.clip(np.exp(-delta_energy), a_max=1.0)
+        return accept_prob, vv_state_new
+
+    def sample_kernel(hmc_state, i):
```

Comment:
@fehiepsi  - I'm not too happy with this unused `i` parameter. Are there other alternatives to `lax.scan` that we can use instead? I'll merge this as is, but was just curious.

Reply:
@neerajprad I think that we'd better remove it. And use `lax.scan` with `lambda i, state: sample_kernel(state)` as in `fori_loop`.